lib/irgen.ml

```
@@ -32,12 +32,35 @@ let lookup (vars : L.llvalue StringMap.t) var =
     raise (Failure (Printf.sprintf "var lookup error: failed to find variable %s\n" var))
 ;;
 
-let build_expr expr vars builder =
+let rec build_expr expr vars builder =
   match expr with
   | SLiteral l -> L.const_int l_int l
   | SBoolLit b -> L.const_int l_bool (if b then 1 else 0)
   | SFloatLit f -> L.const_float l_float f
   | SId var -> L.build_load (lookup vars var) var builder
+  | SEnumAccess (enum_name, variant_name) ->
+    let key = enum_name ^ "::" ^ variant_name in
+    if not (StringMap.mem key vars)
```

technically, we're using `vars` to store variable names, not definitions.
If the enum variant does not exist, the Sast should throw the error here.

So for the IR phase here, we can safely assume that the enum and variant do exist (or fix our sast code).

I do see what you're doing though, mapping the key "index" to the variant, so this is okay at least for now. Maybe we'll find a better way to do it, probably using a new map.

---

Sure, I will leave it here for now till we find a better way.